File: sample-chat-obj-c/sample-chat/Components/UsersInfoScreen/UsersViewController/UserListViewController.m
//
//  UserListViewController.m
//  sample-chat
//
//  Created by Injoit on 30.08.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "UserListViewController.h"
#import "Log.h"
#import "QBUUser+Chat.h"
#import "UIView+Chat.h"
#import "UIViewController+Alert.h"

@interface UserListViewController () <QBChatDelegate>
@end

@implementation UserListViewController

//MARK: - Life Cycle
- (instancetype)initWithNonDisplayedUsers:(NSArray<NSNumber *> *)nonDisplayedUsers {
    if (self = [super initWithNibName:nil bundle:nil]) {
        self.userList = [[UserList alloc] initWithNonDisplayedUsers:nonDisplayedUsers];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UINib *nibUserCell = [UINib nibWithNibName:kUserCellIdentifier bundle:nil];
    [self.tableView registerNib:nibUserCell forCellReuseIdentifier:kUserCellIdentifier];
    self.tableView.allowsMultipleSelection = YES;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.backgroundColor = [UIColor colorWithRed:0.96f green:0.96f blue:0.98f alpha:1.0f];
    self.profile = [[Profile alloc] init];
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self action:@selector(refreshUsers) forControlEvents:UIControlEventValueChanged];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    [self fetchUsers];
}

//MARK - Setup
- (void)fetchUsers {
    [self.refreshControl beginRefreshing];
    __weak __typeof(self)weakSelf = self;
    [self.userList fetchWithPage:1 completion:^(NSArray<QBUUser *> * _Nonnull users, NSError * _Nonnull error) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (users && strongSelf.onFetchedUsers) {
            strongSelf.onFetchedUsers(users);
        }
        [strongSelf.refreshControl endRefreshing];
        [strongSelf.tableView reloadData];
    }];
}

- (void)fetchNext {
    [self.refreshControl beginRefreshing];
    __weak __typeof(self)weakSelf = self;
    [self.userList fetchNextWithCompletion:^(NSArray<QBUUser *> * _Nullable users, NSError * _Nullable error) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (users && strongSelf.onFetchedUsers) {
            strongSelf.onFetchedUsers(users);
        }
        [strongSelf.refreshControl endRefreshing];
        [strongSelf.tableView reloadData];
    }];
}

//MARK: - Actions
- (void)refreshUsers {
    [self fetchUsers];
}

//MARK: - Public Methods
- (void)setupSelectedUsers:(NSArray<QBUUser *> *)users {
    [self.userList appendUsers:users];
    for (QBUUser *user in users) {
        [self.userList.selected addObject:@(user.ID)];
    }
}

- (void)removeSelectedUsers {
    for (NSIndexPath *indexPathForSelectedRow in self.tableView.indexPathsForSelectedRows) {
        [self.tableView deselectRowAtIndexPath:indexPathForSelectedRow animated:NO];
    }
    [self.userList.selected removeAllObjects];
}

- (void)removeSelectedUser:(QBUUser *)user {
    NSUInteger index = [self.userList.fetched indexOfObject:user];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:index inSection:0];
    [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
    [self.userList.selected removeObject:@(user.ID)];
}

- (void)tableView:(UITableView *)tableView
         configureCell:(UserTableViewCell *)cell
     forIndexPath:(NSIndexPath *)indexPath {
    // Can be overridden in a child class.
    NSUInteger lastItemNumber = self.userList.fetched.count - 1;
    if (!self.action && indexPath.row == lastItemNumber && self.userList.isLoadAll == NO) {
        [self fetchNext];
    }
}

#pragma mark - Table view data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.userList.fetched.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 60.0f;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UserTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kUserCellIdentifier];
    if (cell == nil) {
        NSArray *topLevelObjects = [[NSBundle mainBundle] loadNibNamed:kUserCellIdentifier owner:self options:nil];
        cell = [topLevelObjects objectAtIndex:0];
    }
    QBUUser *user = self.userList.fetched[indexPath.row];
    [cell setupUserID:user.ID];
    [cell setupUserName:user.name];
    if (self.profile.ID == user.ID) {
        [cell setupUserName:[NSString stringWithFormat:@"%@ %@", user.name,  @"(You)"]];
    }
    cell.tag = indexPath.row;
    
    [self tableView:tableView configureCell:cell forIndexPath:indexPath];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    QBUUser *user = self.userList.fetched[indexPath.row];
    [self.userList.selected addObject:@(user.ID)];
    if (self.onSelectUser) {
        self.onSelectUser(user, YES);
    }
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath {
    QBUUser *user = self.userList.fetched[indexPath.row];
    if (![self.userList.selected containsObject:@(user.ID)]) {return;}
    [self.userList.selected removeObject:@(user.ID)];
    if (self.onSelectUser) {
        self.onSelectUser(user, NO);
    }
}

- (void)tableView:(UITableView *)tableView willDisplayCell:(nonnull UITableViewCell *)cell forRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    QBUUser *user = self.userList.fetched[indexPath.row];
    if ([self.userList.selected containsObject:@(user.ID)]) {
        [self.tableView selectRowAtIndexPath:indexPath animated:NO scrollPosition:UITableViewScrollPositionNone];
    } else {
        [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
    }
}

#pragma mark - QBChatDelegate
- (void)chatDidConnect {
    [self fetchUsers];
}

- (void)chatDidReconnect {
    [self fetchUsers];;
}

@end

File: sample-chat-swift/sample-chat-swift/Components/Users/ViewControllers/UserListViewController.swift
//
//  UserListViewController.swift
//  sample-chat-swift
//
//  Created by Injoit on 28.12.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox

struct UsersConstant {
    static let perPage:UInt = 100
    static let searchPerPage:UInt = 10
    static let noUsers = "No user with that name"
}

typealias SelectUserCompletion = ( _ user: QBUUser, _ isSelected: Bool) -> Void
typealias FetchedUsersCompletion = ( _ users: [QBUUser]) -> Void

class UserListViewController: UITableViewController {
    //MARK: - Properties
    var action: ChatAction? = nil
    var currentUser = Profile()
    var userList = UserList(nonDisplayedUsers: [])
    var onSelectUser: SelectUserCompletion?
    var onFetchedUsers: FetchedUsersCompletion?
    
    //MARK: - Life Cycle
    init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?, nonDisplayedUsers: [UInt] = []) {
        self.userList = UserList(nonDisplayedUsers: nonDisplayedUsers)
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        tableView.register(UINib(nibName: UserCellConstant.reuseIdentifier, bundle: nil),
                           forCellReuseIdentifier: UserCellConstant.reuseIdentifier)
        tableView.separatorStyle = .none
        tableView.backgroundColor = UIColor(red: 244/255, green: 246/255, blue: 249/255, alpha: 1.0)
        tableView.allowsMultipleSelection = true
        refreshControl = UIRefreshControl()
        refreshControl?.addTarget(self, action: #selector(self.refreshUsers(_:)), for: .valueChanged)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if action != nil {
            return
        }
        fetchUsers()
    }
    
    //MARK - Setup
    func fetchUsers() {
        refreshControl?.beginRefreshing()
        userList.fetchWithPage(1) { [weak self] (users, error) in
            guard let self = self, let users = users else {
                self?.refreshControl?.endRefreshing()
                return
            }
            if users.isEmpty == false {
                self.onFetchedUsers?(users)
            }
            self.refreshControl?.endRefreshing()
            self.tableView.reloadData()
        }
    }
    
    func fetchNext() {
        userList.fetchNext { [weak self] (users, error) in
            guard let self = self, let users = users else {
                return
            }
            if users.isEmpty == false {
                self.onFetchedUsers?(users)
            }
            self.tableView.reloadData()
        }
    }

    //MARK: - Actions
    @objc func refreshUsers(_ sender: UIRefreshControl) {
        fetchUsers()
    }
    
    //MARK: - Public Methods
    func setupSelectedUsers(_ users: [QBUUser]) {
        userList.append(users)
        for user in users {
            userList.selected.insert(user.id)
        }
    }
    
    func removeSelectedUsers() {
        guard let indexPathsForSelectedRows = tableView.indexPathsForSelectedRows else {
            return
        }
        for indexPathForSelectedRow in indexPathsForSelectedRows {
            tableView.deselectRow(at: indexPathForSelectedRow, animated: false)
        }
        userList.selected.removeAll()
    }
    
    func removeSelectedUser(_ user: QBUUser) {
        if let index = userList.fetched.firstIndex(of: user)  {
            let indexPath = IndexPath(row: index, section: 0)
            tableView.deselectRow(at: indexPath, animated: false)
        }
        userList.selected.remove(user.id)
    }
    
    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 60.0
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return userList.fetched.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: UserCellConstant.reuseIdentifier,
                                                       for: indexPath) as? UserTableViewCell else {
            return UITableViewCell()
        }
        let user = userList.fetched[indexPath.row]
        cell.userColor = user.id.generateColor()
        let userName = user.fullName ?? user.login ?? "QB user"
        cell.userNameLabel.text = currentUser.ID == user.id ? userName + " (You)" : userName
        cell.userAvatarLabel.text = String(userName.capitalized.first ?? Character("U"))
        cell.tag = indexPath.row
        
        configure(cell, for: indexPath)
        
        return cell
    }
    
    func configure(_ cell: UserTableViewCell, for indexPath: IndexPath) {
        // Can be overridden in a child class.
        let lastItemNumber = userList.fetched.count - 1
        if action == nil, indexPath.row == lastItemNumber, userList.isLoadAll == false {
            fetchNext()
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let user = userList.fetched[indexPath.row]
        userList.selected.insert(user.id)
        onSelectUser?(user, true)
    }
    
    override func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        let user = userList.fetched[indexPath.row]
        if userList.selected.contains(user.id) == false {
            return
        }
        userList.selected.remove(user.id)
        onSelectUser?(user, false)
    }
    
    override func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        let user = self.userList.fetched[indexPath.row]
        if userList.selected.contains(user.id) {
            tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
            return
        }
        tableView.deselectRow(at: indexPath, animated: false)
    }
}

//MARK: - QBChatDelegate
extension UserListViewController: QBChatDelegate {
    func chatDidConnect() {
        fetchUsers()
    }
    
    func chatDidReconnect() {
        fetchUsers()
    }
}
